rfd/0129-discovery-name-templating.md

```diff
@@ -0,0 +1,370 @@
+
+```yaml
+discovery_service:
+  enabled: true
+  aws:
+    - types: ["rds"]
+      regions: ["us-west-1", "us-west-2"]
+      resource_name_template: "teleport-dev-{{.Name}}-{{.AWS.AccountID}}-{{.AWS.Region}}"
+      tags:
+        "*": "*"
+```
+
+In both the dynamic config and the equivalent static config example, the
+discovery agent will discover AWS RDS databases in the us-west-1 and us-west-2
+AWS regions.
+
+Each database discovered by the `aws` matcher will have the prefix
+"teleport-dev-" followed by the discovered database's name, AWS account ID, and
+AWS region.
+
+For instance, if a database named `foo` is discovered in AWS account ID
+`0123456789012` in region `us-west-1`, the rewritten name will be
+`teleport-dev-foo-012345689012-us-west-1`.
+This disambiguates the database name from other databases in other regions
+and other AWS accounts.
+
+Since the full template syntax is available, a user could modify this to use
+available template builtin functions to shorten the account ID portion:
+`resource_name_template: 'teleport-dev-{{printf "%s-%.4s-%s" .Name .AWS.AccountID .AWS.Region}}'`
+results in: `teleport-dev-foo-0123-us-west-1`
+
+### Available template variables
+
+We must decide which variables to make available for templating.
+
+The supported variables must be sufficient to disambiguate cloud resources to
+avoid name collisions, therefore I think, at a minimum, the supported variables
+should include:
+
+- the original discovered resource name: `.Name`
+- cloud metadata: `.AWS`, `.GCP`, `.Azure`
+  - each of these should only be available for the corresponding matcher
+    type: `aws:`, `gcp:`, `azure:`.
+- the region/location of the resource: `.Region` or `.Location`
+  - this is already included in AWS metadata, but not Azure nor GCP.
+  - rather than a protobuf update, we can just expose this template data
+    variable separately: `.Region` for AWS/Azure and `.Location` for GCP, for
+    consistency with the naming conventions chosen for our discovery matcher
+    config - AWS/Azure matchers use `regions:` and GCP matchers use
+    `locations:`.
+
+For each cloud metadata type, we can make available the corresponding
+`api/types` protobufs:
+
+- `api/types.AWS`
+- `api/types.GCPCloudSQL`
+- `api/types.Azure`
+
+Doing this would couple our cloud metadata protobufs to the supported template
+variables, however it would simplify the implementation and ensure that any new
+cloud metadata is available for templating.
+
+For whatever choice of supported variables, we must document the supported
+template variables in our docs reference.
+
+We should also print a friendly user message that lists the supported
+variables if a user provides a template string that references an unsupported
+variable.
+
+We can check the user provided template when file config is applied,
+or when a dynamic discovery config is created, by executing the template with
+stub data input, or by walking the template parse tree.
+
+### Security
```

Comment:
This looks like a pretty reasonable solution to help reduce the likelihood of conflicts.  However this solution does not appear to make conflicts impossible, or make it so that when a conflict does occur both systems are accessible.

A trusted malicious actor may try to use this ability to deny access to a system (possibly while they conduct an attack on that system and try to prevent others from responding).  That risk is not addressed in this RFD.

That said I still believe this is a valuable feature and does not introduce any new risks.  But long term we should consider how we can have nodes with conflicting names to still be accessible.

Reply:
> This looks like a pretty reasonable solution to help reduce the likelihood of conflicts. However this solution does not appear to make conflicts impossible, or make it so that when a conflict does occur both systems are accessible.

Correct, it does not make conflicts impossible. In fact if you use `resource_name_template: "someConstantName"` then every discovered database will have an identical name. This proposal provides a good heuristic for avoiding name collisions in the common case - where cloud providers allow identical resource names in different accounts/regions. Agreed we should think about addressing the underlying issue of name collisions. 

> But long term we should consider how we can have nodes with conflicting names to still be accessible.

For ssh nodes, the name conflict does not prevent access, because we generate a UUID for nodes and use the hostname as the name displayed in `tsh ls`. It is more tedious to use `tsh ssh` though, since you have to use the uuid instead of the hostname:

```sh
$ tsh ssh gavin@foobar.mars.local.gd                                                                                                                             (base) 
error: ambiguous host could match multiple nodes

Node Name            Node ID                              Address    Labels                             
-------------------- ------------------------------------ ---------- ---------------------------------- 
foobar.mars.local.gd caa749c6-a91b-4530-8849-3f5c08c4db30 ⟵ Tunnel   env=example2,hostname=mac.local.gd 
foobar.mars.local.gd cc76b477-1215-418d-b4cf-6b83597bf02b ⟵ Tunnel   env=example2,hostname=mac.local.gd 

Hint: try addressing the node by unique id (ex: tsh ssh user@node-id)
Hint: use 'tsh ls -v' to list all nodes with their unique ids
```

For resources like databases, name collisions make only one of the databases available at a time. We could do some kind of UUID thing there too? That could be a big change though, and likely would have backward compatibility issues.

Reply:
since we decided against templating, the RFD has been reworked to take a discovery naming convention rather than configurable templates. Therefore, I don't think this security concern applies anymore. I can't think of other concerns